src/sampler.jl

```diff
@@ -88,125 +87,68 @@ function Adaptation.adapt!(
-- The adaptor is set by `adaptor`, for which the default is no adaptation.
-    - It will perform `n_adapts` steps of adaptation, for which the default is the minimum of `1_000` and 10% of `n_samples`
-- `drop_warmup` controls to drop the samples during adaptation phase or not
-- `verbose` controls the verbosity
-- `progress` controls whether to show the progress meter or not
-"""
-function sample(
-    rng::Union{AbstractRNG, AbstractVector{<:AbstractRNG}},
-    h::Hamiltonian,
-    κ::HMCKernel,
-    θ::T,
-    n_samples::Int,
-    adaptor::AbstractAdaptor=NoAdaptation(),
-    n_adapts::Int=min(div(n_samples, 10), 1_000);
-    drop_warmup=false,
-    verbose::Bool=true,
-    progress::Bool=false,
-    (pm_next!)::Function=pm_next!
-) where {T<:AbstractVecOrMat{<:AbstractFloat}}
-    @assert !(drop_warmup && (adaptor isa Adaptation.NoAdaptation)) "Cannot drop warmup samples if there is no adaptation phase."
-    # Prepare containers to store sampling results
-    n_keep = n_samples - (drop_warmup ? n_adapts : 0)
-    θs, stats = Vector{T}(undef, n_keep), Vector{NamedTuple}(undef, n_keep)
-    # Initial sampling
-    h, t = sample_init(rng, h, θ)
-    # Progress meter
-    pm = progress ? ProgressMeter.Progress(n_samples, desc="Sampling", barlen=31) : nothing
-    time = @elapsed for i = 1:n_samples
-        # Make a transition
-        t = transition(rng, h, κ, t.z)
-        # Adapt h and κ; what mutable is the adaptor
-        tstat = stat(t)
-        h, κ, isadapted = adapt!(h, κ, adaptor, i, n_adapts, t.z.θ, tstat.acceptance_rate)
-        tstat = merge(tstat, (is_adapt=isadapted,))
-        # Update progress meter
-        if progress
-            # Do include current iteration and mass matrix
-            pm_next!(pm, (iterations=i, tstat..., mass_matrix=h.metric))
-        # Report finish of adapation
-        elseif verbose && isadapted && i == n_adapts
-            @info "Finished $n_adapts adapation steps" adaptor κ.τ.integrator h.metric
-        end
-        # Store sample
-        if !drop_warmup || i > n_adapts
-            j = i - drop_warmup * n_adapts
-            θs[j], stats[j] = t.z.θ, tstat
-        end
-    end
-    # Report end of sampling
-    if verbose
-        EBFMI_est = EBFMI(map(s -> s.hamiltonian_energy, stats))
-        average_acceptance_rate = mean(map(s -> s.acceptance_rate, stats))
-        if θ isa AbstractVector
-            n_chains = 1
-        else
-            n_chains = size(θ, 2)
-            EBFMI_est = "[" * join(EBFMI_est, ", ") * "]"
-            average_acceptance_rate = "[" * join(average_acceptance_rate, ", ") * "]"
-        end
-        @info "Finished $n_samples sampling steps for $n_chains chains in $time (s)" h κ EBFMI_est average_acceptance_rate
-    end
-    return θs, stats
+    # Compute next transition and state.
+    state = HMCState(0, t, adaptor)
+
+    # Take actual first step
+    return AbstractMCMC.step(rng, model, spl, state; kwargs...)
+end
+
+function AbstractMCMC.step(
```

Comment:
I think we should keep the old signature, which would construct `HamiltonianModel` and `HMCState` then call this function.

Reply:
Hmm, maybe. We have already introduced breaking changes on master with the introduction of `HMCKernel`, etc., so it seems like it's as good a time as any to also transition completely over to AbstractMCMC.jl interface?

Reply:
And it's not like there's anything in the current impl that isn't supported by the new implementation:)

Reply:
There wasn't actually breaking changes to the outmost intefaces (e.g. those need to call sample).

Reply:
Hmm, what do you mean? E.g. the example in the README doesn't work now since `proposal` isn't constructed in the same way (might also be other issues, but that's the immediate one that came to mind).

Reply:
> Hmm, what do you mean? E.g. the example in the README doesn't work now since `proposal` isn't constructed in the same way (might also be other issues, but that's the immediate one that came to mind).

Old constructors should be supported through https://github.com/TuringLang/AdvancedHMC.jl/blob/master/src/AdvancedHMC.jl#L67. 
If there is anything missing we should add them back there.
We should aim to keep outmost interface non-breaking for a few releases before completely removing them.

Also it's a bit cumbersome to write `sample(..., HamiltonianModel(hamiltonian), ..)` instead of sample(..., hamiltonian, ..), no?

Reply:
Sure, I'm cool with adding the old `sample` back and deprecating:) That seems like a clean approach :+1:  

> Also it's a bit cumbersome to write sample(..., HamiltonianModel(hamiltonian), ..) instead of sample(..., hamiltonian, ..), no?

I'm also cool with either of the following:
1. Making `AbstractHamiltonian <: AbstractMCMC.AbstractModel`, which IMO seems sensible.
2. Additional overload that will automagically wrap `hamiltonian` in `HamiltonianModel`.

Reply:
It seems that option 1 is more sensible, as we already making kernel sub-typed of samples from AbstractMCMC.

Reply:
Yeaah I've realized I need to make quite a few changes. Because everything is mutated somehow, we essentially need to put everything in the sampler `state`.

Sooo the question is what we actually use for the `model`, etc. As of right now they would just be "fake" in the sense that they only contain the initial `Hamiltonian` and the initial `AbstractHMCKernel` :confused: We *could* mutate them, but that seems very much against the AbstractMCMC.jl interface...

Reply:
Hmm. Do you mean it actually only make sense to make the log density and gradient as model but treating metric differently?

Reply:
"Only" is maybe strong, but yeah, I think it makes more sense to keep the `hamiltonian` and the `metric` separate. Potentially do away completely with the `hamiltonian` and only keep metric or, alternatively, make `hamiltonian` nothing but a wrapper around `metric` (in case there are future cases we might want to cover) and just make a `DifferentiableDensityModel <: AbstractModel` (could even move this + `DensityModel` from AbstractMH.jl into AbstractMCMC.jl at some point, so these can be re-used e.g. for MALA which is present in AdvancedHMC.jl)